diff --git a/.gitignore b/.gitignore
new file mode 100644
index 0000000..c0a80d6
--- /dev/null
+++ b/.gitignore
@@ -0,0 +1,2 @@
+*.o
+iozone
diff --git a/iozone.c b/iozone.c
index f7da701..3315099 100644
--- a/iozone.c
+++ b/iozone.c
@@ -1838,6 +1838,9 @@ void new_touch_dedup(char *, int);
 /*								*/
 /****************************************************************/
 
+# include <time.h>
+int real_main(int argc, char** argv);
+
 int main(){
 
   char* argvec[7] = { "iozone","-a","-b","results.xls","-+s","-q1M","1024"};
diff --git a/libbif.c b/libbif.c
index b2b9668..8910e48 100644
--- a/libbif.c
+++ b/libbif.c
@@ -17,11 +17,7 @@
 #include <sys/types.h>
 #include <stdio.h>
 #include <sys/file.h>
-#if defined(__AIX__) || defined(__FreeBSD__) || defined(__DragonFly__)
 #include <fcntl.h>
-#else
-#include <sys/fcntl.h>
-#endif
 
 #if defined(OSV5) || defined(linux) || defined (__FreeBSD__) || defined(__OpenBSD__) || defined(__bsdi__) || defined(__APPLE__) || defined(__DragonFly__) || defined(__NetBSD__)
 #include <string.h>
diff --git a/makefile b/makefile
index 518674c..e167b34 100644
--- a/makefile
+++ b/makefile
@@ -160,7 +160,7 @@ generic:	iozone_generic.o  libbif.o
 # Riscv_Musl build with no threads, no largefiles, no async I/O
 #
 riscv_musl:	iozone_riscv_musl.o  libbif.o
-	$(CCRV)  -O $(LDFLAGS) -static iozone_riscv_musl.o libbif.o -o iozone
+	$(CCRV) -g -O $(LDFLAGS) -static iozone_riscv_musl.o libbif.o -o iozone
 
 #
 # No ansii 'C' compiler HP build with no threads, no largefiles, no async I/O
@@ -793,9 +793,9 @@ iozone_riscv_musl.o:	iozone.c libbif.c
 	@echo ""
 	@echo "Building iozone RISC-V + musl-libc (Generic) "
 	@echo ""
-	$(CCRV) -c -O -Dgeneric -Dunix -DHAVE_ANSIC_C -DNO_THREADS \
+	$(CCRV) -g -c -O -Dgeneric -Dunix -DHAVE_ANSIC_C -DNO_THREADS \
 		-DNAME='"RISC-V_musl"' $(CFLAGS) iozone.c -o iozone_riscv_musl.o
-	$(CCRV) -c -O -Dgeneric -Dunix -DHAVE_ANSIC_C -DNO_THREADS \
+	$(CCRV) -g -c -O -Dgeneric -Dunix -DHAVE_ANSIC_C -DNO_THREADS \
 		$(CFLAGS) libbif.c -o libbif.o
 
 
diff --git a/run_iozone.sh b/run_iozone.sh
index 4f03115..32e2b96 100755
--- a/run_iozone.sh
+++ b/run_iozone.sh
@@ -12,13 +12,15 @@ for RUN_N in $(seq $SHORT_REPS); do
     BASE_LOG_FILE=${TEST_LOG_DIR}/base_iozone_${RUN_N}.log
     KEYSTONE_LOG_FILE=${TEST_LOG_DIR}/keystone_iozone_${RUN_N}.log
 
+    echo "Base $RUN_N"
     if [[ $RUN_BASELINE == 1 ]]; then
         ./iozone > ${BASE_LOG_FILE} 2> ${BASE_LOG_FILE}.err
         mv results.xls ${BASE_LOG_FILE}.xls
     fi
 
+    echo "Keystone $RUN_N"
     if [[ $RUN_KEYSTONE == 1 ]]; then
-        ${TEST_RUNNER} iozone ${EYRIE_FULL_SUPPORT} --utm-size ${DEFAULT_USZ} --freemem-size ${LARGE_FSZ} > ${KEYSTONE_LOG_FILE} 2> ${KEYSTONE_LOG_FILE}.err
+        ${TEST_RUNNER} iozone ${EYRIE_FULL_SUPPORT} ${EYRIE_LOADER} --utm-size ${DEFAULT_USZ} --freemem-size ${LARGE_FSZ} > ${KEYSTONE_LOG_FILE} 2> ${KEYSTONE_LOG_FILE}.err
         mv results.xls ${KEYSTONE_LOG_FILE}.xls
     fi
 done
