diff --git a/.gitignore b/.gitignore
new file mode 100644
index 0000000..c0a80d6
--- /dev/null
+++ b/.gitignore
@@ -0,0 +1,2 @@
+*.o
+iozone
diff --git a/iozone.c b/iozone.c
index f7da701..8e96ad5 100644
--- a/iozone.c
+++ b/iozone.c
@@ -1672,7 +1672,7 @@ int client_error;
 
 char pit_hostname[40];
 char pit_service[8];
-int junk;
+static int junk;
 
 /*
  * Host ports used to listen, and handle errors.
@@ -1838,6 +1838,9 @@ void new_touch_dedup(char *, int);
 /*								*/
 /****************************************************************/
 
+# include <time.h>
+int real_main(int argc, char** argv);
+
 int main(){
 
   char* argvec[7] = { "iozone","-a","-b","results.xls","-+s","-q1M","1024"};
diff --git a/libbif.c b/libbif.c
index b2b9668..8e72e63 100644
--- a/libbif.c
+++ b/libbif.c
@@ -55,7 +55,7 @@
 /* Middle Endian */
 #define ENDIAN_4  4
 
-int junk, *junkp;
+static int junk, *junkp;
 
 
 #ifdef HAVE_ANSIC_C
diff --git a/makefile b/makefile
index 518674c..830d19f 100644
--- a/makefile
+++ b/makefile
@@ -157,10 +157,10 @@ generic:	iozone_generic.o  libbif.o
 	$(CC)  -O $(LDFLAGS) iozone_generic.o libbif.o -o iozone
 
 #
-# Riscv_Musl build with no threads, no largefiles, no async I/O
+# RISC-V build with no threads, no largefiles, no async I/O
 #
-riscv_musl:	iozone_riscv_musl.o  libbif.o
-	$(CCRV)  -O $(LDFLAGS) -static iozone_riscv_musl.o libbif.o -o iozone
+riscv:	iozone_riscv.o  libbif.o
+	$(CCRV) -g -O $(LDFLAGS) iozone_riscv.o libbif.o -o iozone
 
 #
 # No ansii 'C' compiler HP build with no threads, no largefiles, no async I/O
@@ -789,13 +789,13 @@ iozone_generic.o:	iozone.c libbif.c
 	$(CC) -c -O -Dgeneric -Dunix -DHAVE_ANSIC_C -DNO_THREADS \
 		$(CFLAGS) libbif.c -o libbif.o
 
-iozone_riscv_musl.o:	iozone.c libbif.c
+iozone_riscv.o:	iozone.c libbif.c
 	@echo ""
-	@echo "Building iozone RISC-V + musl-libc (Generic) "
+	@echo "Building iozone RISC-V (Generic) "
 	@echo ""
-	$(CCRV) -c -O -Dgeneric -Dunix -DHAVE_ANSIC_C -DNO_THREADS \
-		-DNAME='"RISC-V_musl"' $(CFLAGS) iozone.c -o iozone_riscv_musl.o
-	$(CCRV) -c -O -Dgeneric -Dunix -DHAVE_ANSIC_C -DNO_THREADS \
+	$(CCRV) -g -c -O -Dgeneric -Dunix -DHAVE_ANSIC_C -DNO_THREADS \
+		-DNAME='"RISC-V"' $(CFLAGS) iozone.c -o iozone_riscv.o
+	$(CCRV) -g -c -O -Dgeneric -Dunix -DHAVE_ANSIC_C -DNO_THREADS \
 		$(CFLAGS) libbif.c -o libbif.o
 
 
diff --git a/run_iozone.sh b/run_iozone.sh
index 4f03115..32e2b96 100755
--- a/run_iozone.sh
+++ b/run_iozone.sh
@@ -12,13 +12,15 @@ for RUN_N in $(seq $SHORT_REPS); do
     BASE_LOG_FILE=${TEST_LOG_DIR}/base_iozone_${RUN_N}.log
     KEYSTONE_LOG_FILE=${TEST_LOG_DIR}/keystone_iozone_${RUN_N}.log
 
+    echo "Base $RUN_N"
     if [[ $RUN_BASELINE == 1 ]]; then
         ./iozone > ${BASE_LOG_FILE} 2> ${BASE_LOG_FILE}.err
         mv results.xls ${BASE_LOG_FILE}.xls
     fi
 
+    echo "Keystone $RUN_N"
     if [[ $RUN_KEYSTONE == 1 ]]; then
-        ${TEST_RUNNER} iozone ${EYRIE_FULL_SUPPORT} --utm-size ${DEFAULT_USZ} --freemem-size ${LARGE_FSZ} > ${KEYSTONE_LOG_FILE} 2> ${KEYSTONE_LOG_FILE}.err
+        ${TEST_RUNNER} iozone ${EYRIE_FULL_SUPPORT} ${EYRIE_LOADER} --utm-size ${DEFAULT_USZ} --freemem-size ${LARGE_FSZ} > ${KEYSTONE_LOG_FILE} 2> ${KEYSTONE_LOG_FILE}.err
         mv results.xls ${KEYSTONE_LOG_FILE}.xls
     fi
 done
