diff --git a/.gitignore b/.gitignore
index 0641574..e4e6f39 100644
--- a/.gitignore
+++ b/.gitignore
@@ -5,3 +5,4 @@ build/
 riscv/
 riscv64/
 riscv32/
+build-generic64/
diff --git a/mkutils/plat/generic/run.mk b/mkutils/plat/generic/run.mk
index 32f1523..2353373 100644
--- a/mkutils/plat/generic/run.mk
+++ b/mkutils/plat/generic/run.mk
@@ -6,7 +6,7 @@ QEMU_PORT       ?= 9821
 QEMU_DBG_PORT   ?= $(shell echo $$(( $(QEMU_PORT) + 1)) )
 QEMU_DEBUG      := -gdb tcp::$(QEMU_DBG_PORT) -S
 
-QEMU_MEM        ?= 2G
+QEMU_MEM        ?= 8G
 QEMU_SMP        ?= 4
 
 
@@ -16,7 +16,7 @@ QEMU_FLAGS := -m $(QEMU_MEM) -smp $(QEMU_SMP) -nographic \
                 -kernel $(BUILDROOT_BUILDDIR)/images/Image \
                 -drive file=$(BUILDROOT_BUILDDIR)/images/rootfs.ext2,format=raw,id=hd0 \
                 -device virtio-blk-device,drive=hd0 \
-                -append "console=ttyS0 ro root=/dev/vda" \
+                -append "console=ttyS0 ro root=/dev/vda cma=1G" \
                 -netdev user,id=net0,net=192.168.100.1/24,dhcpstart=192.168.100.128,hostfwd=tcp::9821-:22 \
                 -device virtio-net-device,netdev=net0 \
                 -device virtio-rng-pci \
diff --git a/overlays/keystone/configs/linux64-defconfig b/overlays/keystone/configs/linux64-defconfig
index f6e0086..d51fdcb 100644
--- a/overlays/keystone/configs/linux64-defconfig
+++ b/overlays/keystone/configs/linux64-defconfig
@@ -125,3 +125,8 @@ CONFIG_DEBUG_PLIST=y
 CONFIG_DEBUG_SG=y
 # CONFIG_RCU_TRACE is not set
 CONFIG_RCU_EQS_DEBUG=y
+CONFIG_GDB_SCRIPTS=y
+CONFIG_DEBUG_INFO_DWARF4=y
+CONFIG_DEBUG_KERNEL=y
+CONFIG_DEBUG_DRIVER=y
+CONFIG_FRAME_POINTER=y
diff --git a/overlays/keystone/configs/qemu_riscv64_virt_defconfig b/overlays/keystone/configs/qemu_riscv64_virt_defconfig
index 1d348ab..8040b01 100644
--- a/overlays/keystone/configs/qemu_riscv64_virt_defconfig
+++ b/overlays/keystone/configs/qemu_riscv64_virt_defconfig
@@ -30,3 +30,5 @@ BR2_PACKAGE_KEYSTONE_DRIVER=y
 BR2_PACKAGE_HOST_KEYSTONE_SDK=y
 BR2_PACKAGE_KEYSTONE_RUNTIME=y
 BR2_PACKAGE_KEYSTONE_EXAMPLES=y
+BR2_STRIP_strip=n
+BR2_TARGET_ROOTFS_EXT2_SIZE="8G"
