diff --git a/.gitignore b/.gitignore
new file mode 100644
index 0000000..6c18f54
--- /dev/null
+++ b/.gitignore
@@ -0,0 +1,4 @@
+archives
+build
+src
+stamps
diff --git a/bootstrap.sh b/bootstrap.sh
index 66ec3a6..2333f56 100755
--- a/bootstrap.sh
+++ b/bootstrap.sh
@@ -108,6 +108,7 @@ make_directories()
 
 download_prerequisites()
 {
+  set -x
   test -f archives/gmp-${gmp_version}.tar.bz2 || \
       curl -o archives/gmp-${gmp_version}.tar.bz2 \
       https://gmplib.org/download/gmp-${gmp_version}/gmp-${gmp_version}.tar.bz2
